kernel/kernel-core/src/main/java/io/mosip/kernel/core/util/HMACUtils2.java

```
@@ -18,156 +18,185 @@
+    }
+
+    /**
+     * Decodes from BASE64
+     *
+     * @param data data to decode
+     * @return decoded data
+     */
+    public static byte[] decodeBase64(String data) {
+        return BASE64_DECODER.decode(data);
+    }
+
+
+    private static String encode(String password, byte[] salt) {
+        KeySpec spec = new PBEKeySpec(password.toCharArray(), salt, ITERATION_COUNT, 512);
+        try {
+            byte[] key = PBKDF2_FACTORY.get().generateSecret(spec).getEncoded();
+            return encodeBase64String(key);
+        } catch (InvalidKeySpecException e) {
+            throw new RuntimeException("Credential could not be encoded", e);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static String encodeBytesToHex(byte[] byteArray, boolean upperCase, ByteOrder byteOrder) {
+        final int len = byteArray.length;
+
+        // our output size will be exactly 2x byte-array length
+        final char[] buffer = new char[len * 2];
+
+        // choose lower or uppercase lookup table
+        final char[] lookup = upperCase ? LOOKUP_TABLE_UPPER : LOOKUP_TABLE_LOWER;
+
+        int index;
+        for (int i = 0; i < len; i++) {
+            // for little endian we count from last to first
+            index = (byteOrder == ByteOrder.BIG_ENDIAN) ? i : len - i - 1;
+
+            // extract the upper 4 bit and look up char (0-A)
+            buffer[i << 1] = lookup[(byteArray[index] >> 4) & 0xF];
+            // extract the lower 4 bit and look up char (0-A)
+            buffer[(i << 1) + 1] = lookup[(byteArray[index] & 0xF)];
+        }
+        return new String(buffer);
+    }
+
+    private static MessageDigest getDigest(String algo) {
+        try {
+            return MessageDigest.getInstance(algo);
+        } catch (NoSuchAlgorithmException e) {
+            // Should not happen for standard algorithms
+            throw new IllegalStateException(algo + " not supported", e);
```

Can we throw this error :
```
throw new NoSuchAlgorithmException(HMACUtilConstants.MOSIP_NO_SUCH_ALGORITHM_ERROR_CODE.getErrorCode(),
					HMACUtilConstants.MOSIP_NO_SUCH_ALGORITHM_ERROR_CODE.getErrorMessage(), exception.getCause());
```

---

corrected error code